src/dmd/semantic3.d

```diff
@@ -456,8 +456,14 @@ private extern(C++) final class Semantic3Visitor : Visitor
                             stc |= STC.scope_;
                         }
                     }
-                    if (funcdecl.flags & FUNCFLAG.inferScope && !(fparam.storageClass & STC.scope_))
+
+                    // infer scope for lambdas even without -preview=dip1000
+                    // See https://issues.dlang.org/show_bug.cgi?id=20362
+                    const isLambda = funcdecl.isFuncLiteralDeclaration;
+
+                    if ((funcdecl.flags & FUNCFLAG.inferScope || isLambda) && !(fparam.storageClass & STC.scope_))
```

Comment:
Why not set `inferScope` unconditionally for `FuncLiteralDeclaration` instead of special casing things here ?

Reply:
Because it broke a *lot* of things, including dmd's own tests.

Reply:
What I had in mind was: https://github.com/dlang/dmd/pull/10770
Don't see any failure with that.

Reply:
Ah, ok. That's not what I did originally, but even though it was yesterday I can't quite remember what that was exactly.